### PR TITLE
Feature docs

### PR DESCRIPTION
**Pull request by looselycoupled**
This PR is primarily to add comments and update documentation however there are a few minor bug fixes included that were uncovered during documentation.

This is a rather large amount of docs however it should be pretty straightforward to read.  I'll also attach a zip of the HTML that it builds if that helps the review process.

I'd like to call this done for now pending comments, requested changes, etc.  The docs are another case where it can be an unlimited time sink if you let it but I think it's of high enough quality that we can put it out as it stands as long as nothing is incorrect.

@immesys please pay particular attention to the docs surrounding windowing as I feel less confident about some of my explanations.

Once we are both happy enough to merge to the `develop` branch then it will be automatically built and available online at https://btrdb.readthedocs.io/en/develop/ but will not replace the official version at https://btrdb.readthedocs.io/en/latest/.

### CODE REVIEW

**Comment by looselycoupled** [comment]
See attached zip file for the browseable version of the docs.

[build.zip](https://github.com/BTrDB/btrdb4-python/files/2757407/build.zip)

File: btrdb/point.py [hunk 1]
```diff
@@ -23,9 +23,35 @@
...
 
 class RawPoint(object):
+    """
+    A point of data representing a single position within a time series. Each
+    point contains a read-only time and value attribute.
+
+    Parameters
+    ----------
+    time : int
+        The time portion of a single value in the time series.
```
Comment (immesys) [inline btrdb/point.py]: Add: "in nanoseconds since the Unix epoch"

File: btrdb/point.py [hunk 2]
```diff
@@ -23,9 +23,35 @@
...
     def __init__(self, time, value):
-        self.time = time
-        self.value = value
+        self._time = time
+        self._value = value
+
+    @property
+    def time(self):
+        """
+        The time portion of a data point
```
Comment (immesys) [inline btrdb/point.py]: Also add ns since epoch here

File: btrdb/point.py [hunk 3]
```diff
@@ -62,13 +88,52 @@ def __eq__(self, other):
...
+    points of data within a single time series.
+
+    This aggregation point provides for the min, mean, max, count, and standard
+    deviation of all data values it spans.  It is returned by windowing queries
+    such as `windows` or `aligned_windows`.
+
+    Parameters
+    ----------
+    time : int
+        The time in which the aggregated values represent.
```
Comment (immesys) [inline btrdb/point.py]: ns since epoch

File: btrdb/point.py [hunk 4]
```diff
@@ -62,13 +88,52 @@ def __eq__(self, other):
...
+    min : float
+        The minimum value of a time series within a specified range of time.
+    mean : float
+        The mean value of a time series within a specified range of time.
+    max : float
+        The maximum value of a time series within a specified range of time.
+    count : float
+        The number of values of a time series within a specified range of time.
```
Comment (immesys) [inline btrdb/point.py]: s/of/in/

File: btrdb/point.py [hunk 5]
```diff
@@ -62,13 +88,52 @@ def __eq__(self, other):
...
+    stddev : float
+        The standard deviation of the values that this object represents.
```
Comment (immesys) [inline btrdb/point.py]: It is also within the range of time, so maybe have similar doc string

File: btrdb/stream.py [hunk 6]
```diff
@@ -513,9 +548,9 @@ def windows(self, start, end, width, depth=0, version=0):
         end : int
             The end time in nanoseconds for the range to be queried
         width : int
-            Specify the number of ns between data points
+            The number of ns between data points
```
Comment (immesys) [inline btrdb/stream.py]: I would say "The number of nanoseconds in each window, subject to the depth parameter"

File: btrdb/stream.py [hunk 7]
```diff
@@ -513,9 +548,9 @@ def windows(self, start, end, width, depth=0, version=0):
...
         depth : int
-
+            The precision of the data as a power of 2 in nanoseconds.
```
Comment (immesys) [inline btrdb/stream.py]: I would say "The precision of the window duration as a power of 2 in nanoseconds. E.g 30 would make the window duration accurate to roughly 1 second"

File: docs/source/concepts.rst [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+.. note::
+
+	Data requests are fully materialized at this time.  A future release will include the option to process data using generators to save on memory usage.
+
+
+BTrDB Server
+------------
+Like most time series databases, the BTrDB server contains multiple streams of
+data in which each stream contains a data point at a given time.  In BTrDB, only
+one value can be indexed by a given time (we sometimes call this univariate
```
Comment (immesys) [inline docs/source/concepts.rst]: This is maybe misleading, it is possible for a stream to have multiple values with the same timestamp. I get what you are meaning, but the reader may not.

File: docs/source/concepts.rst [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+
+    point.value
+    >> 120.5
+
+StatPoint
+^^^^^^^^^^^^
+The StatPoint provides statistics about multiple points and gives
+aggregation values such as `min`, `max`, `mean`, etc.  This is most useful when you
+don't need to touch every individual value such as when you only need the count
+of the values over a range of time.
```
Comment (immesys) [inline docs/source/concepts.rst]: Also mention that querying stat points executes in time proportional to the number of results, not the number of underlying points (i.e it is logarithmic time)

File: docs/source/working.rst [hunk 10]
```diff
@@ -0,0 +1,14 @@
+Working with btrdb
+====================
+
+Lorem ipsum dolor sit amet, consectetur adipisicing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```
Comment (immesys) [inline docs/source/working.rst]: Hmmm....

File: docs/source/working/stream-view-data.rst [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+
+
+View Windows of Data
+--------------------
+If you don't need to view every single point of data, then it is faster to view
+higher order representations of the data.  BTrDB stores data in
+a tree structure such that the leaves of the tree contain actual values and higher
+nodes store statistical data (min, max, mean, etc.) summaries.  In this schema
+viewing summaries of data involves reading from higher levels of the tree and
+therefore less nodes need to be read from disk.
```
Comment (immesys) [inline docs/source/working/stream-view-data.rst]: Nice

File: docs/source/working/streamsets.rst [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+    >> 4  1500000000400000000                     NaN                     5.0
+    >> 5  1500000000500000000                     6.0                     NaN
+    >> 6  1500000000600000000                     NaN                     7.0
+    >> 7  1500000000700000000                     8.0                     NaN
+    >> 8  1500000000800000000                     NaN                     9.0
+    >> 9  1500000000900000000                    10.0                     NaN
+
+
+Serializing Data
+----------------------------
```
Comment (immesys) [inline docs/source/working/streamsets.rst]: Github formatting is marking this line red, perhaps it doesn't like the length of the underline vs length of the heading

File: docs/source/concepts.rst (hunk 9 above)
Comment (looselycoupled) [reply docs/source/concepts.rst]: replaced with:

These statistical queries execute in time proportional to the number of
results, not the number of underlying points (i.e logarithmic time) and so you
can attain valuable data in a fraction of the time when compared with retrieving
all of the individual values.  Due to the internal data structures, BTrDB does
not need to read the underlying points to return the statistical information!

File: docs/source/concepts.rst (hunk 8 above)
Comment (looselycoupled) [reply docs/source/concepts.rst]: replaced with:

Like most time series databases, the BTrDB server contains multiple streams of
data in which each stream contains a data point at a given time.  However,
BTrDB focuses on univariate data which opens a host of benefits and is one of
the reasons BTrDB is able to process incredibly large amounts of data quickly
and easily.

File: docs/source/working.rst (hunk 10 above)
Comment (looselycoupled) [reply docs/source/working.rst]: oops!  replaced with:

Please review the guided tour linked below to get a better understanding of how
to interact with the BTrDB database.

File: docs/source/working/streamsets.rst (hunk 12 above)
Comment (looselycoupled) [reply docs/source/working/streamsets.rst]: Atom is also not highlighting it correctly in the editor even after deleting and retyping the line, messing with whitespace, etc.  It is rendering correctly as HTML so maybe we live with it if it continues after upcoming commit?

**Comment by looselycoupled** [comment]
@immesys I've implemented all of your comments although I couldnt fix the odd github rendering of the subheading.  I think it might have to do with the dashes used a few lines below in the table.  It does not appear to affect HTML rendering though so I suggest we ignore it for now.

Sorry but I resolved the request changes as I did them not realizing that I would be removing the breadcrumbs for you to review.

Finally, I just asked for permissions to integrate with Travis.  It seems there were multiple hurdles for me to get this done.  Once it's complete I'll have one more commit although I can do that in a separate, tiny PR instead if you prefer.

**Pull request merged**